### ISSUE TITLE
Request change 'concatenate & minifiy' to 'minify & concatenate'

### ISSUE THREAD

**Issue by mkoistinen** [bdkjones/CodeKit-1-Old#20]
I realise this would require more invocations of Uglify, but there's good reason for the change, I think.

When minifying the files independently, each can retain its own comment block at its beginning that contains legal information, copyright info, or just plain a short description of what it is.

When concatenating first, then minifying, only the comment block of the top-most file is retained.

Also, I'm not 100% sure, but there may be some side-effects of concatenating then minifying.

**Comment by bdkjones** [comment]
Hmm... this is an interesting idea.

There is no reason the app can't offer both approaches, although the "minify then concatenate" procedure would be much slower. The reason is a bit technical: the way we hand data off to Uglify.js is by first writing all the JS files to a temporary file on disk, then passing that as a parameter to Uglify.js. (The alternative, which is to simply combine all the JS files into data in RAM and pipe that to Uglify.js using NSPipe hits a very nasty limitation: NSPipe has a max buffer of about 64kb. Once you've sent 64kb into it, you're expected to start reading data from the other side before you can send more through the pipe. Of course, Uglify.js needs ALL the data before it can start chewing, so we can't use this approach because it would fail on any file larger that 64kb.)

My eventual plan is to stop using the binaries that ship with Less, Sass, Stylus, and Uglify. Instead, we'll run each of these directly using JavascriptCore --- the engine at the heart of WebKit. That will eliminate the issue above and we'll be able to simply read the relevant files, construct a Javascript object and run the various compilers on that object. 

This is how JSLint and JSHint currently work within CodeKit because neither of those have a command-line binary. 

The issue is that JavascriptCore is a very low-level API and it's incredibly time-consuming to code for. My plan was to get the app shipping and then migrate languages to JavascriptCore one by one. The only advantage to using JavascriptCore over the current method with NSPipe and NSTask is a bit of a speed increase. But CodeKit will minify the full-blown jQuery library in ~3 seconds on a 2008 MacBook Pro. That's pretty awesome performance.

I will put Uglify.js at the top of the "convert to using JavascriptCore" list and as soon as it does, it will be trivial to offer both approaches.

**Comment by bdkjones** [comment]
After some investigation, I don't believe fulfilling this request is possible.

Part of what uglify.js does to minify files is shorten variable names. If we minify two separate JS files that each begin with a globally-scoped variable, the minified files will have that variable set to "a".

Now, when we throw the two minified scripts into the same file, we have a variable collision: two globally-scoped variables with the same name. Disaster. It's possible to prevent this situation by turning off global variable name obfuscation in the preferences window, but that's way above the average developer's pay grade --- I get tons of emails asking me to allow users to stop the window from coming to the front when there's an error. If people can't figure out THAT preference, there's no hope for something this complicated.

Likewise, I'm not sure that global-namespace collisions are the only ones we'd have to worry about.

Bottom line: Uglify.js is designed to be run ONCE on a FULL script, not multiple times on parts of a script. To do the latter is to invite a huge chance for errors, as the tokenizer in Uglify won't know about all possible tokens.

**Comment by mkoistinen** [comment]
You are correct, of course. All my components are wrapped in closures
though, so I don't think I'd ever see these issues.  I'd hate to be
the one doing tech support for the Great Unwashed with these potential
issues though.

It's a shame we have to eliminate good functionality though because of
idiots. =/

Best regards,
- Martin

100% carbon-neutral since 1-Jan-2008
"Act without activism"

On 19 Sep 2011, at 09:25, bdkjones
reply@reply.github.com
wrote:
